src/aiida/tools/dumping/processes.py

```diff
@@ -0,0 +1,498 @@
+
+        if output_path is None:
+            output_path = self.generate_default_dump_path(process_node=calculation_node)
+
+        try:
+            self.validate_make_dump_path(validate_path=output_path)
+        except:
+            # raise same exception here to communicate it outwards
+            raise
+
+        self.dump_node_yaml(process_node=calculation_node, output_path=output_path)
+
+        io_dump_mapping = self.generate_calculation_io_mapping(io_dump_paths=io_dump_paths)
+
+        # Dump the repository contents of the node
+        calculation_node.base.repository.copy_tree(output_path.resolve() / io_dump_mapping.repository)
+
+        # Dump the extra_inputs
+        if self.include_node_inputs:
+            input_link_manager = calculation_node.base.links.get_incoming(link_type=LinkType.INPUT_CALC)
+            self._dump_calculation_io(parent_path=output_path / io_dump_mapping.inputs, link_manager=input_link_manager)
+
+        # Dump the raw_outputs
+        output_link_manager = calculation_node.base.links.get_outgoing(link_type=LinkType.CREATE)
+        self._dump_calculation_io(
+            parent_path=output_path / io_dump_mapping.outputs,
+            link_manager=output_link_manager,
+        )
+
+    def _dump_calculation_io(self, parent_path: Path, link_manager: LinkManager):
+        for link_triple in link_manager:
+            node = link_triple.node
+            if isinstance(node, FILE_CLASSES) or any(issubclass(type(node), cls) for cls in FILE_CLASSES):
+                file_node_path = self.generate_link_triple_dump_path(
+                    link_triple=link_triple,
+                    parent_path=parent_path,
+                )
+
+                # No .resolve() required as that done in `generate_link_triple_dump_path`
+                link_triple.node.base.repository.copy_tree(file_node_path)
+
+            else:
+                aiida_node_path = self.generate_link_triple_dump_path(
+                    link_triple=link_triple,
+                    parent_path=parent_path / '.aiida_nodes',
+                )
+
+                # This is again QE specific, but, frankly, I don't know how to otherwise separate the pseudos from the
+                # rest of the AiiDA-nodes, as the pseudos are of `Data` type (why not UpfData?), so I cannot distinguish
+                # them from other AiiDA-nodes, such as ArrayData which we definitely want in the hidden `.aiida_nodes`
+                # subdirectory
+                # So if anybody has a better solution, I'd be happy to use that
+                # The problem might be void, though, once all the atomistic code is moved to `aiida-atomistic`
+                if node.node_type == 'data.pseudo.upf.UpfData.':
+                    link_triple.node.base.repository.copy_tree(Path(str(aiida_node_path).replace('.aiida_nodes', '')))
+                elif self.all_aiida_nodes:
+                    link_triple.node.base.repository.copy_tree(aiida_node_path)
```

Comment:
Lots of questions:

* Why use `.aiida_nodes` at all? Why only remove it for pseudos?
* Why are input node repo contents inside `extra_inputs`, but output node repo contents are inside `raw_outputs`?
* How would the problem of pseudos having special treatment be solved if it is moved to `aiida-atomistic`?

What would be wrong with the following folder structure:
```
calculation
├── raw_inputs   # Content of `CalcJobNode` repo
├── raw_outputs  # Content of `CalcJobNode.outputs.retrieved` repo
├── inputs       # Contains repo contents of all input nodes if any
│   └── pseudos
│   |   └── Si
|   |       └── Si.upf
│   └── some_array
│       └── file.npy
│       └── another_file.npy
└── outputs   # Contains repo contents of all output nodes if any, except for `retrieved`

```
To me that seems a lot more coherent. I also think it can make the implementation a whole lot easier and without any exceptions? What am I missing

Reply:
> Why use .aiida_nodes at all? Why only remove it for pseudos?

The idea behind using the hidden `.aiida_nodes` was that in some cases, many such nodes (e.g. `BandsData`, `TrajectoryData`, `ArrayData`, etc.) would be dumped (see the `output_band` and `output_trajectory` folders in the example for part of a `EquationOfStateWorkChain` below). I wouldn't like to immediately expose these to the user once he/she runs `tree` on the dumped directory, which would probably the very first thing they'd do (especially considering the intended user is someone who is not very familiar with AiiDA and wants to look at files instead). 

However, that being said, with the `--all-aiida-nodes` option being turned off by default, the `tree` actually looks fine (see below, as well). So thinking about it now, for me, it would actually be fine to have all of those files in a non-hidden directory. This is then just the effect if the user intentionally specifies `--all-aiida-nodes`.

The problem with the pseudos is that they are stored as plain `Data` nodes. So if I don't explicitly check for them as I'm doing now, they won't be dumped without using `--all-aiida-nodes`, as without that option I'm just dumping `SinglefileData` and `FolderData`, to avoid the cluttering mentioned above. Yet, they should be dumped by default.

**With `--all-aiida-nodes`**

```shell
$ tree -a dump-EquationOfStateWorkChain-816-all-aiida-nodes/02-QuantumEspressoCommonRelaxWorkChain/01-PwRelaxWorkChain/
dump-EquationOfStateWorkChain-816-all-aiida-nodes/02-QuantumEspressoCommonRelaxWorkChain/01-PwRelaxWorkChain
├── .aiida_node_metadata.yaml
└── 01-iteration_01-PwBaseWorkChain
   ├── .aiida_node_metadata.yaml
   ├── 01-create_kpoints_from_distance
   │  ├── .aiida_node_metadata.yaml
   │  └── raw_inputs
   │     └── source_file
   └── 02-iteration_01-PwCalculation
      ├── .aiida_node_metadata.yaml
      ├── extra_inputs
      │  └── pseudos
      │     └── Ag
      │        └── Ag_ONCV_PBE-1.0.oncvpsp.upf
      ├── raw_inputs
      │  ├── .aiida
      │  │  ├── calcinfo.json
      │  │  └── job_tmpl.json
      │  ├── _aiidasubmit.sh
      │  └── aiida.in
      └── raw_outputs
         ├── .aiida_nodes
         │  ├── output_band
         │  │  ├── bands.npy
         │  │  ├── kpoints.npy
         │  │  ├── occupations.npy
         │  │  └── weights.npy
         │  └── output_trajectory
         │     ├── atomic_species_name.npy
         │     ├── cells.npy
         │     ├── energy.npy
         │     ├── energy_accuracy.npy
         │     ├── energy_ewald.npy
         │     ├── energy_hartree.npy
         │     ├── energy_one_electron.npy
         │     ├── energy_smearing.npy
         │     ├── energy_threshold.npy
         │     ├── energy_xc.npy
         │     ├── fermi_energy.npy
         │     ├── forces.npy
         │     ├── positions.npy
         │     ├── scf_accuracy.npy
         │     ├── scf_iterations.npy
         │     ├── steps.npy
         │     ├── stress.npy
         │     └── total_force.npy
         ├── _scheduler-stderr.txt
         ├── _scheduler-stdout.txt
         ├── aiida.out
         └── data-file-schema.xml
```

**Without `--all-aiida-nodes`**

```shell
$ tree -a dump-EquationOfStateWorkChain-816-no-aiida-nodes/02-QuantumEspressoCommonRelaxWorkChain/01-PwRelaxWorkChain/
dump-EquationOfStateWorkChain-816-no-aiida-nodes/02-QuantumEspressoCommonRelaxWorkChain/01-PwRelaxWorkChain
├── .aiida_node_metadata.yaml
└── 01-iteration_01-PwBaseWorkChain
   ├── .aiida_node_metadata.yaml
   ├── 01-create_kpoints_from_distance
   │  ├── .aiida_node_metadata.yaml
   │  └── raw_inputs
   │     └── source_file
   └── 02-iteration_01-PwCalculation
      ├── .aiida_node_metadata.yaml
      ├── extra_inputs
      │  └── pseudos
      │     └── Ag
      │        └── Ag_ONCV_PBE-1.0.oncvpsp.upf
      ├── raw_inputs
      │  ├── .aiida
      │  │  ├── calcinfo.json
      │  │  └── job_tmpl.json
      │  ├── _aiidasubmit.sh
      │  └── aiida.in
      └── raw_outputs
         ├── _scheduler-stderr.txt
         ├── _scheduler-stdout.txt
         ├── aiida.out
         └── data-file-schema.xml
```

> How would the problem of pseudos having special treatment be solved if it is moved to aiida-atomistic?

My idea was that checking could possibly be achieved somewhat more general if the data type comes from an external plugin, but I haven't thought that fully through.

> Why are input node repo contents inside extra_inputs, but output node repo contents are inside raw_outputs?

This is historic. The original idea was to have the `CalculationNode` repo in `raw_inputs`, the linked input node repos in `node_inputs` (or `extra_inputs`), and the `retrieved` output of the `CalculationNode` in `raw_outputs`. After discussion here, I then extended it to not only dump `retrieved`, but all outputs, though, wasn't making the distinction for `retrieved`, so everything just ended up in `raw_outputs` (using `raw_outputs` rather than `extra_outputs` or `node_outputs`, as it contains the actual calculation output file). However, I'm also fine with checking for `retrieved` again specifically and make the distinction as you are proposing.

Reply:
OK, I have implemented it like this. Though, I'd vouch for the following naming:
- `CalculationNode` repository: `inputs`
- `CalculationNode` retrieved output: `outputs`
- Linked node inputs: `node_inputs`
- Linked node outputs (apart from retrieved): `node_outpus`

I think `inputs` and `outputs` is where people would look first and they should contain the files people are used to from running, e.g. QE, classically, so it should have the simplest name.

Reply:
That naming works for me